### PR TITLE
Kylepressel/shoc third moments

### PR DESCRIPTION
**Pull request by pressel**
This PR contains a full refactoring of SHOC's diag_third_shoc_moments subroutine preparing it for porting to C++/Kokkos.

### CODE REVIEW

File: components/cam/src/physics/cam/shoc.F90 [hunk 1]
```diff
@@ -1227,22 +1296,39 @@ subroutine diag_third_shoc_moments(&
...
+    do i=1, shcol 
+
+      tsign = 1._rtype   
       theterm = w_sec_zi(i,k)
       cond = w3clip * sqrt(2._rtype * theterm**3)
       if (w3(i,k) .lt. 0) tsign = -1._rtype
       if (tsign * w3(i,k) .gt. cond) w3(i,k) = tsign * cond
 
-    enddo ! end i loop (column loop)
+    enddo !mend i loop (column loop)
```
Comment (jeff-cohere) [inline components/cam/src/physics/cam/shoc.F90:1328]: Looks like a little typo here.

File: components/cam/src/physics/cam/shoc.F90 [hunk 2]
```diff
@@ -1227,22 +1296,39 @@ subroutine diag_third_shoc_moments(&
...
   enddo ! end k loop (vertical loop)
 
-  return
-
-end subroutine diag_third_shoc_moments
+end subroutine
```
Comment (jeff-cohere) [inline components/cam/src/physics/cam/shoc.F90:1331]: I get leaving off the subroutine name, but maybe we should do this consistently. Not trying to be a pain. :-) We don't have a style guide for this kind of stuff, do we?

File: components/cam/src/physics/cam/shoc.F90 (hunk 1 above)
Comment (pressel) [reply components/cam/src/physics/cam/shoc.F90:1328]: But I mended the loop :) 

Thanks, for catching this. I'll fix it.

File: components/cam/src/physics/cam/shoc.F90 [hunk 3]
```diff
@@ -1147,6 +1137,84 @@ subroutine diag_third_shoc_moments(&
...
+  integer i, j, k, kb, kc
+  real(rtype) :: omega0, omega1, omega2
+  real(rtype) :: X0, Y0, X1, Y1, AA0, AA1
+  real(rtype) :: zvar, x5var, iso, thedz, thedz2
+  real(rtype) :: theterm, cond, tsign
+  real(rtype) :: isosqrt, dthl2, dwthl, dtke, dw2, aw2
+  real(rtype) :: a0, a1, a2, a3, a4, a5
+  real(rtype) :: buoy_sgs2, c, grd, bet2, bet
+  real(rtype) :: f0, f1, f2, f3, f4, f5
   c=7.0_rtype
```
Comment (singhbalwinder) [inline components/cam/src/physics/cam/shoc.F90:1218]: It is great that you brought in these variables in this subroutine as they are used locally used here.  Would it make sense to make them `parameters`? I have been doing that after ensuring that they are indeed parameters. It makes it easier to read the code when we know that these variable will not change during the simulation.

File: components/cam/src/physics/cam/shoc.F90 [hunk 4]
```diff
@@ -1173,6 +1241,7 @@ subroutine diag_third_shoc_moments(&
       isosqrt=iso**2
       buoy_sgs2=isosqrt*brunt_zi(i,k)
       bet2=ggr/thetal_zi(i,k)
+
```
Comment (singhbalwinder) [inline components/cam/src/physics/cam/shoc.F90:1244]: Nitpick: I think we can move `kb` and `kc` computation just after `do k=2,nlev` as we do not need to recompute `kb` and `kc` for each `i`.

File: components/cam/src/physics/cam/shoc.F90 (hunk 2 above)
Comment (pressel) [reply components/cam/src/physics/cam/shoc.F90]: That wasn't intentional. I have fixed it.

File: components/cam/src/physics/cam/shoc.F90 (hunk 3 above)
Comment (pressel) [reply components/cam/src/physics/cam/shoc.F90]: Done, good suggestion. Thanks.

**Pull request merged**